### ISSUE TITLE
Every realm.write increases number of active versions and never decreases.

### ISSUE THREAD

**Issue by KimiChiu** [realm/realm-cocoa#7068]
## Goals
Decreasing the number of active versions.

## Expected Results
maximumNumberOfActiveVersions = 20 won't crash the app.

## Actual Results
App crashes after calling `insert` multiple times on the main thread.

## Steps for others to Reproduce
Call `insert` multiple times.

## Code Sample
   ```
 func insert(aid: String, boardTitle: String, title: String,
                category: String, author: String, date: String,
                articleHeader: String, owner: String) throws -> ArticleLog? {
        logHelper.d("insert")

        return try autoreleasepool {
            let realm = try Realm()
            let now = NSDate()

            let data = ArticleLog.newInstance(
                    id: getWorkerThreadID(realm),
                    aid: aid,
                    boardTitle: boardTitle,
                    title: title,
                    category: category,
                    author: author,
                    date: date,
                    articleHeader: articleHeader,
                    pushed: false,
                    owner: owner,
                    createdDate: now,
                    lastUpdateDate: now)

            try realm.write {
                realm.add(data)
            }

            return data
        }
    }
```

`getWorkerThreadID` is just a method to get the id.

```
    func getWorkerThreadID(_ realm: Realm) -> Int {
        logHelper.d("getWorkerThreadID")
        semaphore.wait()
        defer {
            semaphore.signal()
        }
        _id =  autoIncrement(
                realm: realm,
                id: _id,
                realmObject: ENTITY_CLASS,
                primaryKey: ArticleLog.Column.PrimaryKey.rawValue)
        return _id;
    }

    internal func autoIncrement(realm: Realm, id: Int, realmObject: Object.Type, primaryKey: String) -> Int{
        if( id == -1 ){
            return ( realm.objects( realmObject ).max(ofProperty: primaryKey) as Int? ?? 0 ) + 1
        }
        
        return id + 1
    }
```
## Version of Realm and Tooling

Realm framework version: 10.1.4

Realm Object Server version: N/A

Xcode version: 12.3

iOS/OSX version: 14.3

Dependency manager + version: CocoaPods 1.10.0

Things I already tried:

1. Remove `autoreleasepool`.
2. Make it return nothing.
3. Call `insert` in `DispatchQueue.main.async`.

I have no idea why the number of active versions won't be decreased.

**Comment by ericjordanmossman** [comment]
@KimiChiu Do you loop over insert multiple times when you call it from the main thread?

**Comment by KimiChiu** [comment]
@ericjordanmossman I found the problem. I have a background thread looper which is calling `realm.objects(ENTITY_CLASS).filter`.
It's nothing about the `insert` method.
When I debug with memory graph, I found there were 5 realmResults rather than 5 realm instances.
But I only call `realm.objects(ENTITY_CLASS).filter` once in the test and the results were never be stored in any global variable.
Why did these results exist in memory?

```
    if let filter = ArticleKeywordFilterService.shared.find(realm, by: b) {
        articleKeywordFilters[filter.board] = filter.keywords.split(on: ",", omitEmptyStrings: true, trimResults: true)
    }
```

I'm now using `autoreleasepool` to release everything in this scope.
But there's a `Fatal Exception: std::__1::system_error
mutex lock failed: Invalid argument` when I'm using `autoreleasepool`.
It's not reproducible on my test device.
Is there a way to release everything without using `autoreleasepool`?
